File: src/pratice/backtraking/N과M15649.java
package pratice.backtraking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;


public class N과M15649 {
    public static int[] selected;
    public static boolean[] used;
    public static int N, M;
    public static StringBuffer buffer = new StringBuffer();
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String[] split = reader.readLine().split(" ");
        N = Integer.parseInt(split[0]);
        M = Integer.parseInt(split[1]);
        selected = new int[M + 1];
        used = new boolean[N + 1];
        function(0);
        System.out.println(buffer);
    }

    public static void function(int curIdx){
        if(curIdx >= M){
            for (int i = 0; i < curIdx; i++) {
                buffer.append(selected[i] + " ");
            }
            buffer.append("\n");
        }else{
            for (int i = 1; i <=N ; i++) {
                if(!used[i]) {
                    selected[curIdx] = i;
                    used[i]=true;
                    function(curIdx + 1);
                    selected[curIdx] = 0;
                    used[i]=false;
                }
            }
        }
    }
}

File: src/pratice/backtraking/N과M15649.kt
package pratice.backtraking

private var used = BooleanArray(0)
private var selected = IntArray(0)
private val buffer: StringBuffer = StringBuffer()
private var N = 0
private var M = 0

fun main(){
    val num = readln().split(" ").map { n -> n.toInt() }.toList()
    N = num[0]
    M = num[1]
    used = BooleanArray(N + 1)
    selected = IntArray(M + 1)
    recFunc(1)
    println(buffer)
}


private fun recFunc(k: Int) {
    if (k >= M + 1){
        for(i in 1 until selected.size){
            buffer.append("${selected[i]} ")
        }
        buffer.append("\n")
    }else {
        for(i in 1 .. N){
            if(!used[i]){
                selected[k] = i
                used[i] = true
                recFunc(k + 1)
                used[i] = false
                selected[k] = 0
            }
        }
    }
}
